zerver/tests/test_email_mirror.py

```diff
@@ -596,13 +610,194 @@ def test_receive_stream_email_include_quotes_success(self) -> None:
+        self.assertEqual(bot.bot_owner.id, hamlet.id)
+        result = self.client_get(f"/json/streams/{channel.id}/email_address", {"sender_id": bot.id})
+        self.assert_json_success(result)
+
+        # Sender is a random user ID. (None of the above three cases)
+        othello = self.example_user("othello")
+        result = self.client_get(
+            f"/json/streams/{channel.id}/email_address", {"sender_id": othello.id}
+        )
+        self.assert_json_error(
+            result,
+            "The sender ID must be either the current user's ID, the email gateway bot's ID, or the ID of a bot owned by the user.",
+        )
+
+    def test_creator_with_send_message_permission(self) -> None:
+        hamlet = self.example_user("hamlet")
+        realm = get_realm("zulip")
+        channel = get_stream("Denmark", realm)
+
+        do_change_user_role(hamlet, UserProfile.ROLE_MODERATOR, acting_user=None)
+        do_change_stream_post_policy(
+            channel, Stream.STREAM_POST_POLICY_MODERATORS, acting_user=hamlet
+        )
+
+        # Sender is the current user itself.
+        channel_email_address = encode_email_address(channel, creator=hamlet, sender=hamlet)
+        incoming_valid_message = self.create_incoming_valid_message(channel_email_address)
+
+        with self.assertLogs(logger_name, level="INFO") as m:
+            process_message(incoming_valid_message)
+        self.assertEqual(
+            m.output,
+            [
+                f"INFO:{logger_name}:Successfully processed email to {channel.name} ({realm.string_id})"
+            ],
+        )
+
+        # Sender is the Email gateway bot.
+        email_gateway_bot = get_system_bot(settings.EMAIL_GATEWAY_BOT, realm.id)
+        channel_email_address = encode_email_address(
+            channel, creator=hamlet, sender=email_gateway_bot
+        )
+        incoming_valid_message = self.create_incoming_valid_message(channel_email_address)
+
+        with self.assertLogs(logger_name, level="INFO") as m:
+            process_message(incoming_valid_message)
+        self.assertEqual(
+            m.output,
+            [
+                f"INFO:{logger_name}:Successfully processed email to {channel.name} ({realm.string_id})"
+            ],
+        )
+
+        # Sender is a bot owned by the current user + has NO post permission.
+        bot = self.create_test_bot("test2", hamlet, full_name="Test bot")
+        assert bot.bot_owner is not None
+        self.assertEqual(bot.bot_owner.id, hamlet.id)
+        channel_email_address = encode_email_address(channel, creator=hamlet, sender=bot)
+        incoming_valid_message = self.create_incoming_valid_message(channel_email_address)
+
+        with self.assertLogs(logger_name, level="INFO") as m:
+            process_message(incoming_valid_message)
+        self.assertEqual(
+            m.output,
+            [
+                f"INFO:{logger_name}:Successfully processed email to {channel.name} ({realm.string_id})"
```

Comment:
This case feels a bit strange that the bot doesn't have the permission to post but still it'd appear as a sender because the owner has the permission. -- Our core `access_stream_for_send_message` function allows to post message in such a case, is it an expected behaviour ?

Reply:
Yeah, that's the current security model; I don't see why the bot sending via an email would feel any different from it doing it directly via the API. We will probably change that security model eventually as we add more convenient permissions configuration options.